### PR TITLE
Adding basic optimization algorithm that matches a single order against an AMM

### PR DESCRIPTION
**Pull request by harisang**
Still need to implement a token conservation test, so as to make sure that rounding errors do not create any issues. Moreover, the generated output is still incomplete. Planning to resolve this soon.

### CODE REVIEW

File: src/api/aux.cpp [hunk 1]
```diff
@@ -22,6 +22,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
     
     for (auto &i: json_file["tokens"].items()) {
         boost::multiprecision::mpf_float t;
+        bool normalize_priority;
+
+        normalize_priority = (i.value()["normalize_priority"] == 1) ? true : false;
```
Comment (marcovc) [inline src/api/aux.cpp]: this may be simplified to:
```c++
normalize_priority = i.value()["normalize_priority"] == 1;
```

File: src/solver/optimize.cpp [hunk 2]
```diff
@@ -1,14 +1,294 @@
...
+#include <sstream> 
 #include "../components/token.hpp"
 #include "../components/order.hpp"
 #include "../components/amm.hpp"
 #include "./optimize.hpp"
+#include "../../external_libs/json.hpp"
 
-void solve_auction(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms)
+
+// auxiliary function (used temporarily) for computing x^n using the boost library data types
```
Comment (marcovc) [inline src/solver/optimize.cpp]: [It seems](https://www.boost.org/doc/libs/1_78_0/libs/multiprecision/doc/html/boost_multiprecision/ref/number.html) there is a power function in the library. Is there any reason why you're coding your own?

File: src/solver/optimize.cpp [hunk 3]
```diff
@@ -1,14 +1,294 @@
...
+        if ((x0 <= 0) || (y0 <= 0))
+            continue;
+
+        // at this point, we have found a "matching" AMM
+        // we must satisfy x1 * y1 >= x0 * y0, where (x1,y1) is the new pool balance and (x0, y0) is the original pool balance
+        // Since it is a fill-or-kill sell order, we have x1 = x0 + 0.997y_
+        // and y1 = y0 - x
+        // Thus, maximally exploiting the pool, we get that y1 = x0 * y0 / x1,
+        // which implies that x = y0 * (1 - x0/x1)
+        rational_x1 = static_cast<boost::multiprecision::mpf_float>(x0) + 0.997 * static_cast<boost::multiprecision::mpf_float>(y_);
```
Comment (marcovc) [inline src/solver/optimize.cpp]: Where is the 0.997 coming from?

File: src/solver/optimize.cpp [hunk 4]
```diff
@@ -1,14 +1,294 @@
...
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
+    boost::multiprecision::mpf_float rational_x1;
```
Comment (marcovc) [inline src/solver/optimize.cpp]: I am a bit confused by the name, is this mpf_float a float or a rational?

**Review by marcovc — COMMENTED** [review]
Thanks for the PR. Just some comments to get started, I'll do a more thorough review next week.

File: src/api/aux.cpp (hunk 1 above)
Comment (harisang) [reply src/api/aux.cpp]: Done

File: src/solver/optimize.cpp (hunk 3 above)
Comment (harisang) [reply src/solver/optimize.cpp]: Oh sorry, my bad. I didn't realize I had hardwired this number in the code. I replaced it with (1-fee) which should be the right thing to do for a Uniswap Constant Product AMM.

File: src/solver/optimize.cpp (hunk 4 above)
Comment (harisang) [reply src/solver/optimize.cpp]: Sorry for the confusion. The name just meant that this is a float version of the mpz_int variable x1. Bad choice of name, I agree. I fixed it.

File: src/solver/optimize.cpp (hunk 2 above)
Comment (harisang) [reply src/solver/optimize.cpp]: Good point. Will look into it. I had done this as a temporary fix and forgot to replace it.

File: src/api/aux.cpp [hunk 5]
```diff
@@ -22,6 +22,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
     
     for (auto &i: json_file["tokens"].items()) {
         boost::multiprecision::mpf_float t;
```
Comment (marcovc) [inline src/api/aux.cpp]: Sugestion: rename "t" as external_price, or "ep", or something that is easy to remember.

File: src/api/aux.cpp [hunk 6]
```diff
@@ -30,9 +33,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
             t = static_cast<boost::multiprecision::mpf_float>(tt);
```
Comment (marcovc) [inline src/api/aux.cpp]: Does this casting work?

File: src/api/aux.cpp [hunk 7]
```diff
@@ -30,9 +33,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
         }
         if (i.value()["alias"].is_null())
-            tokens.push_back(Token(i.key(), "_NULL", i.value()["decimals"], t));       // TODO: Check if multiple tokens with "null" alias can create any issue
+            tokens.push_back(Token(i.key(), "_NULL", i.value()["decimals"], t, normalize_priority));       // TODO: Check if multiple tokens with "null" alias can create any issue
         else
-            tokens.push_back(Token(i.key(), i.value()["alias"], i.value()["decimals"], t));
+            tokens.push_back(Token(i.key(), i.value()["alias"], i.value()["decimals"], t, normalize_priority));
```
Comment (marcovc) [inline src/api/aux.cpp]: On an if-then-else, it is good practice (and usually used by the compiler) to write the more likely branch first.

File: src/api/aux.cpp [hunk 8]
```diff
@@ -22,6 +22,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
     
     for (auto &i: json_file["tokens"].items()) {
```
Comment (marcovc) [inline src/api/aux.cpp]: Suggestion: rename "i" as "token", or "t" or something that is easy to remember.

File: src/api/aux.cpp [hunk 9]
```diff
@@ -30,9 +33,9 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
             t = static_cast<boost::multiprecision::mpf_float>(tt);
         }
         if (i.value()["alias"].is_null())
-            tokens.push_back(Token(i.key(), "_NULL", i.value()["decimals"], t));       // TODO: Check if multiple tokens with "null" alias can create any issue
+            tokens.push_back(Token(i.key(), "_NULL", i.value()["decimals"], t, normalize_priority));       // TODO: Check if multiple tokens with "null" alias can create any issue
         else
-            tokens.push_back(Token(i.key(), i.value()["alias"], i.value()["decimals"], t));
+            tokens.push_back(Token(i.key(), i.value()["alias"], i.value()["decimals"], t, normalize_priority));
         _idx_tokens[i.key()] = _num_tokens-1;
     }
```
Comment (marcovc) [inline src/api/aux.cpp:48]: Suggestion: rename "i" as "order" or "o" or something that is easy to remember.

Comment (marcovc) [reply src/api/aux.cpp:48]: Same for all t_i variables in the loop.

Comment (marcovc) [reply src/api/aux.cpp:48]: and tt_i

File: src/solver/optimize.cpp [hunk 10]
```diff
@@ -1,14 +1,296 @@
...
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
```
Comment (marcovc) [inline src/solver/optimize.cpp:18]: Suggestion x = buy_amount, y = sell_amount (so that the reader of the code doesn't have to read the paper).

File: src/solver/optimize.cpp [hunk 11]
```diff
@@ -1,14 +1,296 @@
...
+        res = res * x;
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
```
Comment (marcovc) [inline src/solver/optimize.cpp:17]: Could we pass order[i] instead of i as a parameter, or do we really need to pass the i ?

File: src/solver/optimize.cpp [hunk 12]
```diff
@@ -1,14 +1,296 @@
...
+
     
+    x_ = orders[i].buy_amount;
+    y_ = orders[i].sell_amount;
+    num_amms = amms.size();
+
+    for (j = 0; j < num_amms; j++) {
+        x0 = 0;
+        y0 = 0;
+        if (amms[j].sell_token_index == orders[i].sell_token_index)
```
Comment (marcovc) [inline src/solver/optimize.cpp:32]: If we add some "else" 's here then code will run faster.

File: src/solver/optimize.cpp [hunk 13]
```diff
@@ -1,14 +1,296 @@
...
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
```
Comment (marcovc) [inline src/solver/optimize.cpp]: In c++ we usually declare the variables as close as possible to their use. For example x0 could be declared at line 43.

File: src/solver/optimize.cpp [hunk 14]
```diff
@@ -1,14 +1,296 @@
...
+        if (amms[j].sell_token_index == orders[i].sell_token_index)
+            x0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].sell_token_index)
+            x0 = amms[j].buy_reserve_amount;
+        if (amms[j].sell_token_index == orders[i].buy_token_index)
+            y0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].buy_token_index)
+            y0 = amms[j].buy_reserve_amount;
+
+        if ((x0 <= 0) || (y0 <= 0))
```
Comment (marcovc) [inline src/solver/optimize.cpp]: tip: no need for parenthesis, and can use "or" instead of "||".

File: src/solver/optimize.cpp [hunk 15]
```diff
@@ -1,14 +1,296 @@
...
+            }
+            else {
+                best_order.sell_token_price = 1.0;
+                best_order.buy_token_price = (static_cast<boost::multiprecision::mpf_float>(y_)) / (static_cast<boost::multiprecision::mpf_float>(x));
+            }
+            best_order.buy_token_price = best_order.buy_token_price * pow(10, tokens[orders[i].buy_token_index].decimals);
+            best_order.sell_token_price = best_order.sell_token_price * pow(10, tokens[orders[i].sell_token_index].decimals);
+        }
+    }
+    return;
```
Comment (marcovc) [inline src/solver/optimize.cpp:86]: No need for this return.

File: src/solver/optimize.cpp [hunk 16]
```diff
@@ -1,14 +1,296 @@
...
+
+
+void generate_output(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, ExecutedOrder &best_order, nlohmann::json &output)
+{
+    bool found;
+    boost::multiprecision::mpz_int temp;
+
+    found = (best_order.objective_value > 0) ? true : false;
+
+    output["amms"];
```
Comment (marcovc) [inline src/solver/optimize.cpp:176]: Interesting, I suppose this is here so that the "amms" key is created in the output object, right?

**Review by marcovc — COMMENTED** [review]
Here's another c++ only review. Will get to the logical part afterwards!

File: src/api/aux.cpp (hunk 5 above)
Comment (harisang) [reply src/api/aux.cpp]: Done

File: src/api/aux.cpp (hunk 8 above)
Comment (harisang) [reply src/api/aux.cpp]: Done

File: src/solver/optimize.cpp (hunk 12 above)
Comment (harisang) [reply src/solver/optimize.cpp:32]: Partially addressed this. I am not a huge fan of the nested "else" since indentation creates some issues with readability, but you are right in terms of the speed-up.

File: src/solver/optimize.cpp (hunk 14 above)
Comment (harisang) [reply src/solver/optimize.cpp]: Coming from C, I am a bit more comfortable with "||" and slightly prefer it to be honest! I did remove the nested parentheses.

File: src/solver/optimize.cpp (hunk 15 above)
Comment (harisang) [reply src/solver/optimize.cpp:86]: Although not needed, I tend to prefer it, as for me it is an explicit way of stating that the function is over. It has been a bit of a habit of mine for quite some time, but if you think that it might confuse people, I can remove it.

File: src/solver/optimize.cpp (hunk 16 above)
Comment (harisang) [reply src/solver/optimize.cpp:176]: Yes, that's correct. Not sure if this is the "correct" way to do so, though. Will look into it again, because I agree that it looks a bit strange.

File: src/solver/optimize.cpp (hunk 13 above)
Comment (harisang) [reply src/solver/optimize.cpp]: I suspect that the compiler in any case takes care of all these, but isn't technically "more correct" to define these variables up top since they are reused throughout all the iterations of the for-loop? Again, I suspect the compiler would do the right thing and not dynamically allocate the variables at the beginning of each iteration, but this is a more explicit way of doing so. I don't have a strong opinion on that so I could change it of course

File: src/solver/optimize.cpp (hunk 12 above)
Comment (marcovc) [reply src/solver/optimize.cpp:32]: I usually ident it as:
```c++
if (cond1)
   // do something
else
if (cond2)
  // do something else
else
  // do another thing
```

File: src/solver/optimize.cpp (hunk 13 above)
Comment (marcovc) [reply src/solver/optimize.cpp]: You're right, the compiler will likely move it around, but remember that in c++ creating and destroying an object also runs code. So the place where you declare the object is important, as the constructor and destructor will run when it enters/leaves scope. Another reason is that it is easier to see what is the type of an object if it is declared closer to the place where it is actually used. Finally, this is the more idiomatic way of coding c++.

File: src/solver/optimize.cpp (hunk 16 above)
Comment (marcovc) [reply src/solver/optimize.cpp:176]: I think that works! The way I usually do is something in the direction of 
```c++
output.push("amms"); // not sure if "push" is the name of the function we want
```

File: src/api/aux.cpp [hunk 17]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-            t = -1;
-        else {
-            //std::string tt = std::to_string(i.value()["external_price"]);
-            std::string tt = (i.value()["external_price"]).dump();
-            t = static_cast<boost::multiprecision::mpf_float>(tt);
+    for (auto &tkn: json_file["tokens"].items()) {
+        boost::multiprecision::mpf_float external_price;
+        bool normalize_priority;
+
+        normalize_priority = (tkn.value()["normalize_priority"] == 1);
```
Comment (marcovc) [inline src/api/aux.cpp]: Since normalize_priority is a constant, it is more explicit if we do:
```c++
bool const normalize_priority = (tkn.value()["normalize_priority"] == 1);
```

File: src/api/aux.cpp [hunk 18]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
+        if (!tkn.value()["external_price"].is_null()) {
+            std::string external_price_string = (tkn.value()["external_price"]).dump();
```
Comment (marcovc) [inline src/api/aux.cpp]: also a const

File: src/api/aux.cpp [hunk 19]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-
-        orders.push_back(Order(t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11,t12));
+    for (auto &o: json_file["orders"].items()) {
+        int key = std::stoi(o.key());
+        int sell_token = _idx_tokens.at(o.value()["sell_token"]);
+        int buy_token = _idx_tokens.at(o.value()["buy_token"]);
+        bool is_sell_order = o.value()["is_sell_order"];
+        bool is_liquidity_order = o.value()["is_liquidity_order"];
+        bool allow_partial_fill = o.value()["allow_partial_fill"];
+        std::string sell_amount_string = o.value()["sell_amount"];
```
Comment (marcovc) [inline src/api/aux.cpp]: const

File: src/api/aux.cpp [hunk 20]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
+        boost::multiprecision::mpz_int cost_amount = static_cast<boost::multiprecision::mpz_int>(cost_amount_string);
+
+        orders.push_back(Order(key, sell_token, buy_token, is_sell_order, is_liquidity_order, allow_partial_fill, sell_amount, buy_amount, fee_token, fee_amount, cost_token, cost_amount));
     }
 
     // For-loop that reads all AMM information
-    for (auto &i: json_file["amms"].items()) {
-        int t1 = std::stoi(i.key());
+    for (auto &am: json_file["amms"].items()) {
+        int key = std::stoi(am.key());
```
Comment (marcovc) [inline src/api/aux.cpp]: watch out that the id of orders/amms can be an arbitrary string

File: src/api/aux.cpp [hunk 21]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-        boost::multiprecision::mpz_int t8 = static_cast<boost::multiprecision::mpz_int>(tt8);
-        int t9 = _idx_tokens.at(i.value()["fee"]["token"]);
-        std::string tt10 = i.value()["fee"]["amount"];
-        boost::multiprecision::mpz_int t10 = static_cast<boost::multiprecision::mpz_int>(tt10);
-        int t11 = _idx_tokens.at(i.value()["cost"]["token"]);
-        std::string tt12 = i.value()["cost"]["amount"];
-        boost::multiprecision::mpz_int t12 = static_cast<boost::multiprecision::mpz_int>(tt12);
-
-        orders.push_back(Order(t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11,t12));
+    for (auto &o: json_file["orders"].items()) {
```
Comment (marcovc) [inline src/api/aux.cpp]: Since we're not modifying "o" below, we should declare it as "auto const& o".

File: src/api/aux.cpp [hunk 22]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
+        std::string cost_amount_string = o.value()["cost"]["amount"];
+        boost::multiprecision::mpz_int cost_amount = static_cast<boost::multiprecision::mpz_int>(cost_amount_string);
+
+        orders.push_back(Order(key, sell_token, buy_token, is_sell_order, is_liquidity_order, allow_partial_fill, sell_amount, buy_amount, fee_token, fee_amount, cost_token, cost_amount));
     }
 
     // For-loop that reads all AMM information
-    for (auto &i: json_file["amms"].items()) {
-        int t1 = std::stoi(i.key());
+    for (auto &am: json_file["amms"].items()) {
```
Comment (marcovc) [inline src/api/aux.cpp]: auto const& am

File: src/api/aux.cpp [hunk 23]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-        boost::multiprecision::mpz_int t3[2];
+        int pools_idx[2];
+        boost::multiprecision::mpz_int pools_reserves[2];
         int counter = 0;
 
-        for (auto &j: i.value()["reserves"].items()) {
-            t2[counter] = _idx_tokens.at(j.key());
-            std::string tt3 = j.value();
-            t3[counter] = static_cast<boost::multiprecision::mpz_int>(tt3);
+        for (auto &j: am.value()["reserves"].items()) {
```
Comment (marcovc) [inline src/api/aux.cpp]: can we have a more descriptive name for j?

File: src/api/aux.cpp [hunk 24]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-    for (auto &i: json_file["tokens"].items()) {
-        boost::multiprecision::mpf_float t;
-        if (i.value()["external_price"].is_null())
-            t = -1;
-        else {
-            //std::string tt = std::to_string(i.value()["external_price"]);
-            std::string tt = (i.value()["external_price"]).dump();
-            t = static_cast<boost::multiprecision::mpf_float>(tt);
+    for (auto &tkn: json_file["tokens"].items()) {
+        boost::multiprecision::mpf_float external_price;
```
Comment (marcovc) [inline src/api/aux.cpp:33]: tip: do something like
```c++
typedef boost::multiprecision::mpf_float BigFloat;
typedef boost::multiprecision::mpz_int BigInt;
```

somewhere at the top of the file (or in a file you always include), and then use BigFloat/BigInt instead of the larger type identifiers.

File: src/api/aux.cpp [hunk 25]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
+    for (auto &o: json_file["orders"].items()) {
+        int key = std::stoi(o.key());
+        int sell_token = _idx_tokens.at(o.value()["sell_token"]);
+        int buy_token = _idx_tokens.at(o.value()["buy_token"]);
+        bool is_sell_order = o.value()["is_sell_order"];
+        bool is_liquidity_order = o.value()["is_liquidity_order"];
+        bool allow_partial_fill = o.value()["allow_partial_fill"];
+        std::string sell_amount_string = o.value()["sell_amount"];
+        boost::multiprecision::mpz_int sell_amount = static_cast<boost::multiprecision::mpz_int>(sell_amount_string);
+        std::string buy_amount_string = o.value()["buy_amount"];
```
Comment (marcovc) [inline src/api/aux.cpp]: This pattern is used in so many places, that you could consider to factor it out into a `parse_int_value(std::string const& key)` function.

File: src/api/aux.cpp [hunk 26]
```diff
@@ -19,72 +19,75 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
         // We will only use ConstantProduct AMMs, so this check ensures that we ignore the rest
-        if (i.value()["kind"] != "ConstantProduct")
+        if (am.value()["kind"] != "ConstantProduct")
             continue;
         
-        int t2[2];
-        boost::multiprecision::mpz_int t3[2];
+        int pools_idx[2];
+        boost::multiprecision::mpz_int pools_reserves[2];
         int counter = 0;
```
Comment (marcovc) [inline src/api/aux.cpp]: Since "counter" can't be negative, better options are `unsigned int counter;` or even better `std::size_t counter` (which is usually used for counters).

**Review by marcovc — COMMENTED** [review]
Looking better! More c++ reviews.

File: src/solver/optimize.cpp (hunk 12 above)
Comment (harisang) [reply src/solver/optimize.cpp:32]: Yes, you have a point. I always felt that this style was a bit ambiguous and was using brackets instead, but I will probably reconsider now and follow your suggestion.

File: src/api/aux.cpp (hunk 17 above)
Comment (harisang) [reply src/api/aux.cpp]: Fixed it. Thanks, such comments help a lot since I am not used to the "const" keyword yet.

File: src/api/aux.cpp (hunk 24 above)
Comment (harisang) [reply src/api/aux.cpp:33]: I generally avoid too many definitions as I think that then it gets harder to actually keep track of the definitions themselves. Of course, so far, I haven't used any, so I will follow your advice!

File: src/solver/optimize.cpp (hunk 2 above)
Comment (harisang) [reply src/solver/optimize.cpp]: Seems that I was actually not using after some point. For now, the standard pow() function of C++ is used. Will test if that is enough or not.

File: src/api/aux.cpp (hunk 6 above)
Comment (harisang) [reply src/api/aux.cpp]: Seems that it does. Didn't think this through too much, but I ended up using this instead of the C-style cast I had in mind, i.e., `t = (boost::multiprecision::mpf_float) tt;`

File: src/api/aux.cpp [hunk 27]
```diff
@@ -19,72 +29,65 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
+        int const cost_token = _idx_tokens.at(o.value()["cost"]["token"]);
+        boost::multiprecision::mpz_int const cost_amount = parse_string_to_mpz_int(o.value()["cost"]["amount"]);
+
+        orders.push_back(Order(key, sell_token, buy_token, is_sell_order, is_liquidity_order, allow_partial_fill, sell_amount, buy_amount, fee_token, fee_amount, cost_token, cost_amount));
     }
 
     // For-loop that reads all AMM information
-    for (auto &i: json_file["amms"].items()) {
-        int t1 = std::stoi(i.key());
+    for (auto const& am: json_file["amms"].items()) {
```
Comment (josojo) [inline src/api/aux.cpp:68]: Maybe amm is a better name?

File: src/api/aux.cpp [hunk 28]
```diff
@@ -19,72 +29,65 @@ void parse_json_file(std::vector<Token> &tokens, std::vector<Order> &orders, std
...
-        if (i.value()["external_price"].is_null())
-            t = -1;
-        else {
-            //std::string tt = std::to_string(i.value()["external_price"]);
-            std::string tt = (i.value()["external_price"]).dump();
-            t = static_cast<boost::multiprecision::mpf_float>(tt);
-        }
-        if (i.value()["alias"].is_null())
-            tokens.push_back(Token(i.key(), "_NULL", i.value()["decimals"], t));       // TODO: Check if multiple tokens with "null" alias can create any issue
+    for (auto const& tkn: json_file["tokens"].items()) {
```
Comment (josojo) [inline src/api/aux.cpp:32]: I think writing token instead of tkn is clearer and will it make easier to read for others.

File: src/solver/main_solver.cpp [hunk 29]
```diff
@@ -40,18 +40,18 @@ int main()
...
     
-    // Just printing out the input
     std::cout.precision(std::numeric_limits<boost::multiprecision::mpf_float>::digits10);
 
-    for (auto &i: tokens)
-        print_token(i);
 
-    for (auto &i: orders)
-        print_order(i, tokens);
+    //for (auto &i: tokens)
```
Comment (josojo) [inline src/solver/main_solver.cpp:46]: The best way to deal with this is to have some debugging mode and only print it in debugging. But this should be done in another pr

File: src/solver/optimize.cpp (hunk 10 above)
Comment (josojo) [reply src/solver/optimize.cpp:18]: Yes, please :)

**Review by josojo — COMMENTED** [review]
Good start. Generally, it makes sense to me. But its hard to verify that the code works without any unit test.

Small tip for the next time: You could have just put the function `process_sell_order` into the PR, get all the feedback on it, add unit test for it, and trying to apply the best practices for that part. 

Then the next PR for adding `process_buy_order` is much easier, for you and the reviewers.

**Comment by harisang** [comment]
> Good start. Generally, it makes sense to me. But its hard to verify that the code works without any unit test.
> 
> Small tip for the next time: You could have just put the function `process_sell_order` into the PR, get all the feedback on it, add unit test for it, and trying to apply the best practices for that part.
> 
> Then the next PR for adding `process_buy_order` is much easier, for you and the reviewers.

I fully agree with you. As is, it is hard to follow everything. My bad, and will make sure to add some tests.

**Comment by nlordell** [comment]
Please reopen https://github.com/cowprotocol/simple-solver-template if needed.
